adafruit_requests.py

```diff
@@ -54,22 +54,40 @@
 __version__ = "0.0.0-auto.0"
 __repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Requests.git"
 
-_the_interface = None  # pylint: disable=invalid-name
-_the_sock = None  # pylint: disable=invalid-name
+# This module mirror CPython's socket class. It is settable because the network devices are external
+# to CircuitPython.
+_socket = None  # pylint: disable=invalid-name
 
 
-def set_socket(sock, iface=None):
-    """Helper to set the global socket and optionally set the global network interface.
-    :param sock: socket object.
-    :param iface: internet interface object
+def set_socket(socket):
+    """Helper to set the global socket.
+    :param sock: socket module
 
     """
-    global _the_sock  # pylint: disable=invalid-name, global-statement
-    _the_sock = sock
-    if iface:
-        global _the_interface # pylint: disable=invalid-name, global-statement
-        _the_interface = iface
-        _the_sock.set_interface(iface)
+    global _socket  # pylint: disable=invalid-name, global-statement
+    _socket = socket
+
+# Hang onto open sockets so that we can reuse them.
+_socket_pool = {} # pylint: disable=invalid-name
+def _get_socket(host, port, proto, *, timeout=1):
+    key = (host, port, proto)
+    if key in _socket_pool:
+        socket = _socket_pool[key]
+        if not socket.connected():
+            del _socket_pool[key]
+        else:
+            return socket
+    addr_info = _socket.getaddrinfo(host, port, 0, _socket.SOCK_STREAM)[0]
+    sock = _socket.socket(addr_info[0], addr_info[1], addr_info[2])
+    sock.settimeout(timeout)  # socket read timeout
+
+    if proto == "https:":
+        # for SSL we need to know the host name
+        sock.connect((host, port), _socket.TLS_MODE)
+    else:
+        sock.connect(addr_info[-1], _socket.TCP_MODE)
```

Comment:
@tannewt  
```
Traceback (most recent call last):
  File "code.py", line 46, in <module>
  File "/lib/adafruit_requests.py", line 340, in get
  File "/lib/adafruit_requests.py", line 300, in request
  File "/lib/adafruit_requests.py", line 89, in _get_socket
AttributeError: 'module' object has no attribute 'TCP_MODE'
```

`_socket` does not have a _socket.TLS_MODE/TCP_MODE attribute, the attribute is defined within the sockets interface as `socket._the_interface.TLS_MODE`

Would we want to define the TLS/TCP modes in ESP32SPI?

Reply:
That is part of the change I have pending. We can wait for this PR until the ESP32SPI changes are in.